squeal-postgresql/src/Squeal/PostgreSQL/Query.hs

```diff
@@ -623,6 +640,185 @@ offset
   -> TableExpression schema params relations grouping
 offset off rels = rels {offsetClause = off : offsetClause rels}
 
+{-----------------------------------------
+JSON stuff
+-----------------------------------------}
+
+type PGjson_each_variant val tab =
+  '[ tab ::: '[ "key" ::: 'NotNull 'PGtext, "value" ::: 'NotNull val ] ]
+
+type PGjson_each tab = PGjson_each_variant 'PGjson tab
+type PGjsonb_each tab = PGjson_each_variant 'PGjsonb tab
+type PGjson_each_text tab = PGjson_each_variant 'PGtext tab
+type PGjsonb_each_text tab = PGjson_each_variant 'PGtext tab
+
+unsafeAliasedFromClauseExpression
+  :: Aliased (Expression schema' relations' grouping' params') ty'
+  -> FromClause schema params relations
+unsafeAliasedFromClauseExpression aliasedExpr = UnsafeFromClause
+  (renderAliasedAs renderExpression aliasedExpr)
+
+-- | Expands the outermost JSON object into a set of key/value pairs.
+jsonEach
```

Comment:
Given a JSON object like `{ "key": value }` it produces a table `(key string, value jsonb)`;

```
psql=# select t.* from json_each('{"x":3,"y":{"z":12}}') as t;
 key |  value   
-----+----------
 x   | 3
 y   | {"z":12}

```

Reply:
Ok, I take this to mean that `json_each` can only be applied to single values, that is it cannot refer to any columns in any tables. In this case it's closed with respect to the `relations` type parameter. There are two ways of representing this, with a frankentype or the empty list.

```haskell
(forall relations. Expression schema relations 'Ungrouped params)
```
or
```haskell
Expression schema '[] 'Ungrouped params
```

I'd pick the latter.

Reply:
I don't think Squeal can support such queries because `FromClause`s don't have a relations parameter which would be necessary for them to reference columns in another relation previously defined. Look at it this way, you have right now a type signature:

```Haskell
jsonEach
  :: Aliased (Expression schema relations 'Ungrouped params) '(tab, nullity 'PGjson)
  -> FromClause schema params (PGjson_each tab)
```

The `relations` parameter gets lost. It might be a later feature to add one into from clauses, but it doesn't seem that useful.

Also thanks for adding json literals! That's super useful I think.